log4j-core/src/test/java/org/apache/logging/log4j/core/util/WatchHttpTest.java

```diff
@@ -0,0 +1,175 @@
+import static com.github.tomakehurst.wiremock.client.WireMock.stubFor;
+import static com.github.tomakehurst.wiremock.client.WireMock.urlPathEqualTo;
+import static com.github.tomakehurst.wiremock.core.WireMockConfiguration.wireMockConfig;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+
+/**
+ * Test the WatchManager
+ */
+public class WatchHttpTest {
+
+    private static final String FORCE_RUN_KEY = WatchHttpTest.class.getSimpleName() + ".forceRun";
+    private final String file = "log4j-test1.xml";
+    private static FastDateFormat formatter;
+    private static final String XML = "application/xml";
+
+    private static final boolean IS_WINDOWS = PropertiesUtil.getProperties().isOsWindows();
+
+    @BeforeClass
+    public static void beforeClass() {
+        try {
+            formatter = FastDateFormat.getInstance("EEE, dd MMM yyyy HH:mm:ss", TimeZone.getTimeZone("UTC"));
+        } catch (Exception ex) {
+            System.err.println("Unable to create date format.");
+            ex.printStackTrace();
+            throw ex;
+        }
+    }
+
+    @Rule
+    public WireMockRule wireMockRule = new WireMockRule(wireMockConfig().dynamicPort().dynamicHttpsPort()
+        .keystorePath(TestConstants.KEYSTORE_FILE)
+        .keystorePassword(String.valueOf(TestConstants.KEYSTORE_PWD()))
+        .keystoreType(TestConstants.KEYSTORE_TYPE));
+
+    @Test
+    public void testWatchManager() throws Exception {
+        BlockingQueue<String> queue = new LinkedBlockingQueue<>();
+        List<ConfigurationListener> listeners = new ArrayList<>();
+        listeners.add(new TestConfigurationListener(queue, "log4j-test1.xml"));
+        TimeZone timeZone = TimeZone.getTimeZone("UTC");
+        Calendar now = Calendar.getInstance(timeZone);
+        Calendar previous = now;
+        previous.add(Calendar.MINUTE, -5);
+        Configuration configuration = new DefaultConfiguration();
+        Assume.assumeTrue(!IS_WINDOWS || Boolean.getBoolean(FORCE_RUN_KEY));
+        URL url = new URL("http://localhost:" + wireMockRule.port() + "/log4j-test1.xml");
+        StubMapping stubMapping = stubFor(get(urlPathEqualTo("/log4j-test1.xml"))
+            .willReturn(aResponse()
+            .withBodyFile(file)
+            .withStatus(200)
+            .withHeader("Last-Modified", formatter.format(previous) + " GMT")
+            .withHeader("Content-Type", XML)));
+        final ConfigurationScheduler scheduler = new ConfigurationScheduler();
+        scheduler.incrementScheduledItems();
+        final WatchManager watchManager = new WatchManager(scheduler);
+        watchManager.setIntervalSeconds(1);
+        scheduler.start();
+        watchManager.start();
+        try {
+            watchManager.watch(new Source(url.toURI(), previous.getTimeInMillis()), new HttpWatcher(configuration, null,
+                listeners, previous.getTimeInMillis()));
+            Thread.sleep(1500);
```

Comment:
Is there any way you could instead expose a `CoundDownLatch` for testing purposes? Then you could `await()` it instead of allowing any problems under CI load.

Reply:
Turns out I didn't need to sleep at all anyway...